rust/src/protocol/checkpoints.rs

```diff
@@ -141,148 +140,55 @@ pub async fn create_checkpoint_for(
-    candidates.sort_by(|a, b| a.0.cmp(&b.0));
-
-    let mut last_file: (i64, ObjectMeta) = (
-        0,
-        ObjectMeta {
-            location: Path::from(""),
-            last_modified: DateTime::<Utc>::MIN_UTC,
-            size: 0,
-            e_tag: None,
-        },
-    );
-    let file_needs_time_adjustment =
-        |current_file: &(i64, ObjectMeta), last_file: &(i64, ObjectMeta)| {
-            last_file.0 < current_file.0
-                && last_file.1.last_modified.timestamp() >= current_file.1.last_modified.timestamp()
-        };
-
-    let should_delete_file = |file: &(i64, ObjectMeta)| {
-        file.1.last_modified.timestamp() <= log_retention_timestamp && file.0 < until_version
-    };
-
-    let mut maybe_delete_files: Vec<(i64, ObjectMeta)> = Vec::new();
-    let mut files_to_delete: Vec<(i64, ObjectMeta)> = Vec::new();
-
-    // Init
-    if !candidates.is_empty() {
-        let removed = candidates.remove(0);
-        last_file = (removed.0, removed.1.clone());
-        maybe_delete_files.push(removed);
-    }
-
-    let mut current_file: (i64, ObjectMeta);
-    loop {
-        if candidates.is_empty() {
-            deleted_log_num += flush_delete_files(
-                storage,
-                &mut maybe_delete_files,
-                &mut files_to_delete,
-                should_delete_file,
-            )
-            .await?;
-
-            return Ok(deleted_log_num);
-        }
-        current_file = candidates.remove(0);
-
-        if file_needs_time_adjustment(&current_file, &last_file) {
-            let updated = (
-                current_file.0,
-                ObjectMeta {
-                    location: current_file.1.location.clone(),
-                    last_modified: last_file.1.last_modified.add(Duration::seconds(1)),
-                    size: 0,
-                    e_tag: None,
-                },
-            );
-            maybe_delete_files.push(updated);
-            last_file = (
-                maybe_delete_files.last().unwrap().0,
-                maybe_delete_files.last().unwrap().1.clone(),
-            );
-        } else {
-            let deleted = flush_delete_files(
-                storage,
-                &mut maybe_delete_files,
-                &mut files_to_delete,
-                should_delete_file,
-            )
-            .await?;
-            if deleted == 0 {
-                return Ok(deleted_log_num);
-            }
-            deleted_log_num += deleted;
+    // Feed a stream of candidate deletion files directly into the delete_stream
+    // function to try to improve the speed of cleanup and reduce the need for
+    // intermediate memory.
+    let deleted = storage
+        .delete_stream(
+            storage
+                .list(Some(storage.log_path()))
+                .await?
+                // Pass along only the Ok results from storage.list
+                .filter(|res| futures::future::ready(res.is_ok()))
```

Comment:
This was the only reason I didn't take this approach first. If we are ok with not handling the error here that's fine with me!

Reply:
Given the nature of the function, I don't think we would want to halt execution on a single error would we? 

Assuming that's correct, I I will modify the code to log errors but continue processing